### PR TITLE
scx_layered: Add timer helpers

### PR DESCRIPTION
**Pull request by hodgesds**
Add a registry of timers and a timer manager for running bpf timers.

### CODE REVIEW

**Comment by hodgesds** [comment]
Tested with (since CI is broken):
```
$ cargo build --release && sudo ./target/release/scx_layered --stats 1 --run-example
   Compiling scx_rustland v1.0.5 (/home/hodgesd/scx/scheds/rust/scx_rustland)
   Compiling scx_rlfifo v1.0.5 (/home/hodgesd/scx/scheds/rust/scx_rlfifo)
    Finished `release` profile [optimized] target(s) in 4.65s
14:42:27 [INFO] CPUs: online/possible=80/80 nr_cores=40
14:42:27 [INFO] Topology awareness not specified, selecting enabled based on hardware
14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_init not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_exit not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_prep_move not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_move not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_cancel_move not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] libbpf: struct_ops layered: member cgroup_set_weight not found in kernel, skipping it as it's set to zero

14:42:27 [INFO] Layered Scheduler Attached. Run `scx_layered --monitor` for metrics.
###### Tue, 22 Oct 2024 07:42:28 -0700 ######
tot=      6 local=83.33 open_idle=50.00 affn_viol= 0.00 proc=2ms nodes=2
busy= 33.3 util=    4.5 load=      4.8 fallback_cpu=  7
excl_coll=0.00 excl_preempt=0.00 excl_idle=0.00 excl_wakeup=0.00
  batch    : util/dcycle/frac/  1.4/ 30.7/   30.7 load/load_frac_adj/frac=     0.77/30.70/ 16.0 tasks=  1070
             tot=      2 local=100.0 wake/exp/reenq= 0.00/ 0.00/ 0.00
             xlayer_wake= 0.00 xlayer_rewake= 0.00
             keep/max/busy= 0.00/ 0.00/ 0.00 kick= 0.00 yield/ign= 0.00/    0 
             open_idle= 0.00 mig= 0.00 xnuma_mig= 0.00 xllc_mig= 0.00 affn_viol= 0.00
             preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec=100.0/   1.97ms, slice=20ms
             cpus=  6 [  0,  6] 00000013 00001300 00000000
  immediate: util/dcycle/frac/  0.0/  0.0/    0.0 load/load_frac_adj/frac=     0.00/0.00/  0.0 tasks=     0
             tot=      0 local= 0.00 wake/exp/reenq= 0.00/ 0.00/ 0.00
             xlayer_wake= 0.00 xlayer_rewake= 0.00
             keep/max/busy= 0.00/ 0.00/ 0.00 kick= 0.00 yield/ign= 0.00/    0 
             open_idle= 0.00 mig= 0.00 xnuma_mig= 0.00 xllc_mig= 0.00 affn_viol= 0.00
             preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=20ms
             cpus= 66 [  0, 72] ffffff80 ffff80ff 0000ffff
             excl_coll= 0.00 excl_preempt= 0.00
  normal   : util/dcycle/frac/  3.1/ 69.3/   69.3 load/load_frac_adj/frac=     4.04/69.30/ 84.0 tasks=   409
             tot=      4 local=75.00 wake/exp/reenq=25.00/ 0.00/ 0.00
             xlayer_wake= 0.00 xlayer_rewake= 0.00
             keep/max/busy= 0.00/ 0.00/ 0.00 kick= 0.00 yield/ign= 0.00/    0 
             open_idle=75.00 mig= 0.00 xnuma_mig= 0.00 xllc_mig= 0.00 affn_viol= 0.00
             preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec=525.0/   4.10ms, slice=20ms
             cpus=  8 [  0,  8] 0000006c 00006c00 00000000
  stress-ng: util/dcycle/frac/  0.0/  0.0/    0.0 load/load_frac_adj/frac=     0.00/0.00/  0.0 tasks=     0
             tot=      0 local= 0.00 wake/exp/reenq= 0.00/ 0.00/ 0.00
             xlayer_wake= 0.00 xlayer_rewake= 0.00
             keep/max/busy= 0.00/ 0.00/ 0.00 kick= 0.00 yield/ign= 0.00/    0 
             open_idle= 0.00 mig= 0.00 xnuma_mig= 0.00 xllc_mig= 0.00 affn_viol= 0.00
             preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=0.8ms
             cpus=  0 [  0,  0] 00000000 00000000 00000000
^CEXIT: Scheduler unregistered from user space
```

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 1]
```diff
@@ -49,6 +50,29 @@ static u32 preempt_cursor;
...
+}
+
+static __always_inline bool run_timer_cb(int key)
+{
+	switch (key) {
+	case NOOP_TIMER:
+	case MAX_TIMERS:
+	default:
+		return noop_timer();
+		break;
```
Comment (JakeHillion) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:66]:
```suggestion
		return false;
```

**Review by JakeHillion — APPROVED** [review]

**Comment by likewhatevs** [comment]
> I'm not a fan of run_timer_cb having to be defined in the same translation unit as the rest of the stuff in timer.bpf.c, but I as discussed offline I don't see anything we can do to fix that. The include ordering requirements are not good, but this does seem better than having timer maps scattered everywhere.

fwiw I'd have said scattered maps are better than splitting things like this because this splitting makes it harder for tooling to work well and like, tooling says this and it looks like it's right to me:

WezTerm 2024-10-24 10 33 01

**Comment by likewhatevs** [comment]
(will fix that in a pr after one that makes tooling work the best it can w/ this kind of splitting, but ideally just don't split things like this unless we have to due to bpf weirdness etc.)

**Comment by JakeHillion** [comment]
> > I'm not a fan of run_timer_cb having to be defined in the same translation unit as the rest of the stuff in timer.bpf.c, but I as discussed offline I don't see anything we can do to fix that. The include ordering requirements are not good, but this does seem better than having timer maps scattered everywhere.
> 
> fwiw I'd have said scattered maps are a much better approach than splitting things like this because this splitting makes it harder for tooling to work well and like, tooling says this and it looks like it's right to me:
> 
> WezTerm 2024-10-24 10 33 01

Isn't this totally right? `15 * NSEC_PER_SEC` and `15*10^9 >> 2^31`, your tooling seems to be doing well

**Comment by likewhatevs** [comment]
> Isn't this totally right? 15 * NSEC_PER_SEC and 15*10^9 >> 2^31, your tooling seems to be doing well

Ahh yeah that's my point, like, I marginally know what I'm doing at best but yellow and red squigglies etc. help me make stuff that will maybe compile at least lol.

Timers being in their own file(s) instead of being lumped into main.bpf.c limit what tooling can like, grok/check etc. (I think this is the best that is possible, and those red's should be like, not red):

WezTerm 2024-10-24 10 52 42

**Comment by JakeHillion** [comment]
> > Isn't this totally right? 15 * NSEC_PER_SEC and 15*10^9 >> 2^31, your tooling seems to be doing well
> 
> Ahh yeah that's my point, like, I marginally know what I'm doing at best but yellow and red squigglies etc. help me make stuff that will maybe compile at least lol.
> 
> Timers being in their own file(s) instead of being lumped into main.bpf.c limit what tooling can like, grok/check etc. (I think this is the best that is possible, and those red's should be like, not red):
> 
> WezTerm 2024-10-24 10 52 42

This as well is just a bug. That file should be written in such a way that it can compile by itself, and those red warnings means it needs a couple of forward decls. This isn't an issue with the tooling or the idea of splitting files, it's an issue with the method we're doing for it - included them as text. This hides the issue for people who don't use the plugins because the file is never compiled alone.

Basically it's a code quality thing. We can add the correct forward decls for `layered_timers` and `run_timer_cb` to the header which will make the warnings go away.

**Comment by likewhatevs** [comment]
> This isn't an issue with the tooling or the idea of splitting files, it's an issue with the method we're doing for it - included them as text.

Kinda disagree, if no split everything work out the gate.

> We can add the correct forward decls for layered_timers and run_timer_cb to the header which will make the warnings go away.

Yeah but that's another thing that can drift vs just using a few big files and everything working (until underlying tooling better supports splitting things etc.).

**Comment by JakeHillion** [comment]
> > This isn't an issue with the tooling or the idea of splitting files, it's an issue with the method we're doing for it - included them as text.
> 
> Kinda disagree, if no split everything work out the gate.

I guess we have different intentions with devx, but for me small files are a million times easier to work with than a big one. My brain doesn't work well in big files, so not everything ;)

> > We can add the correct forward decls for layered_timers and run_timer_cb to the header which will make the warnings go away.
> 
> Yeah but that's another thing that can drift vs just using a few big files and everything working (until underlying tooling better supports splitting things etc.).

This is some work that can be done, though I don't know enough about compiling `.bpf.c` files to do it. If every `.bpf.c` file compiles fine with clang then it should also run fine under clangd - the issues that showed up here were because we don't compile the file in isolation, we only compile it as a textual include in `main.bpf.c`.

If we add "compile and discard" for each of these files in the BPF builder it should become something that's CI checked and won't break. We could even force registration of `.bpf.c` files by messing with include paths, which would make remembering to add them required.
